tests/interlocked_linux_ut/interlocked_linux_ut.c

```diff
@@ -0,0 +1,710 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#ifdef __cplusplus
+#include <cstdlib>
+#include <cstddef>
+#include <cstdatomic>
+#else
+#include <stdlib.h>
+#include <stddef.h>
+#include <stdbool.h>
+#include <stdatomic.h>
+#endif
+#include "azure_macro_utils/macro_utils.h"
+#include "testrunnerswitcher.h"
+
+#include "umock_c/umock_c.h"
+#include "umock_c/umocktypes_stdint.h"
+
+#include "azure_c_util/interlocked.h"
+
+
+static TEST_MUTEX_HANDLE g_testByTest;
+
+MU_DEFINE_ENUM_STRINGS(UMOCK_C_ERROR_CODE, UMOCK_C_ERROR_CODE_VALUES)
+
+static void on_umock_c_error(UMOCK_C_ERROR_CODE error_code)
+{
+    ASSERT_FAIL("umock_c reported error :%" PRI_MU_ENUM "", MU_ENUM_VALUE(UMOCK_C_ERROR_CODE, error_code));
+}
+
+#define ENABLE_MOCKS
+#include "mock_interlocked.h"
+#undef ENABLE_MOCKS
+
+static int16_t expected_atomic_compare_exchange_16_value;
+static bool hook_mock_atomic_compare_exchange_16(volatile_atomic int16_t* object, int16_t* expected, int16_t desired)
+{
+    ASSERT_ARE_EQUAL(int16_t, expected_atomic_compare_exchange_16_value, *expected);
+    int16_t original_expected = *expected;
+    *expected = *object;
+    return original_expected == *object;
```

Comment:
Actually, maybe this is easier to just implement as a "real" here for the compare exchange mocks.

Leave the ASSERT, then replace these three lines with `return atomic_compare_exchange(object, expected, desired);`

You don't rename the real functions in the test, so this will call the real atomic functions without mocks, but it's called within a mock call for all the nice call recording/mocking functionality.